### ISSUE TITLE
The specified cast from a materialized type is not valid.

### ISSUE THREAD

**Issue by Bladixx** [zzzprojects/EntityFramework-Plus#135]
I have this query to my entity:

```
Repository.AsQueryable()
                    .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted))
                    .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted).Select(t => t.Customer))
                    .IncludeFilter(x => x.Customer)
                    .FirstOrDefault(x => x.Id == id && !x.Deleted);
```

the query  works but after some time i start getting this exception:
 

> System.InvalidOperationException
> The specified cast from a materialized 'System.Collections.Generic.IEnumerable`1[Data.Entities.Project]' type to the 'Data.Entities.Project' type is not valid.

I note that after recycle the appPool the query starts working again.

My entities are:

 ```
public class Project 
    {
        public int Id { get; set; }

        public string Name { get; set; }

        public int? CustomerId { get; set; }
     
        public int? ParentId { get; set; }

        public virtual List<ProjectBudget> Budget { get; set; }

        public virtual Customer Customer { get; set; }

        public virtual List<Project> ChildProjects { get; set; } = new List<Project>();

        public virtual Project Parent { get; set; }
    }

      public class Customer : 
    {
        public int Id { get; set; }

        public string Name { get; set;  }
        public string Description { get; set; }

        public virtual ICollection<Project> Projects { get; set; }
    }

```
Thanks

**Comment by JonathanMagnan** [comment]
Hello @Bladixx ,

I don't believe this issue is caused directly by this feature.

Are you using another third party library for Entity Framework or other features from our library?

By example, are you using [EF+ Query Filter](https://github.com/zzzprojects/EntityFramework-Plus/wiki/EF-Query-Filter-%7C-Entity-Framework-Dynamic-Instance-and-Global-Filters)? There is more chance this issue is caused by a side effect from this feature.

Could you provide the full stack trace? Perhaps that could help us to find out what's happening.

Best Regards,

Jonathan

**Comment by Bladixx** [comment]
Hello @JonathanMagnan

No I'm not using another third party library for EF. From your library at the moment I'm only using Z.EntityFramework.Plus.EF6, most specifically the extension IncludeFilter() without any Query Filter.

My stack trace is:

```
System.InvalidOperationException: The specified cast from a materialized 'System.Collections.Generic.IEnumerable`1[Data.Entities.Project]' type to the 'Data.Entities.Project' type is not valid.
   at System.Data.Entity.Core.Common.Internal.Materialization.CodeGenEmitter.CheckedConvert[TSource,TTarget](TSource value)
   at System.Data.Entity.Core.Common.Internal.Materialization.Coordinator`1.ReadNextElement(Shaper shaper)
   at System.Data.Entity.Core.Common.Internal.Materialization.Shaper`1.RowNestedResultEnumerator.MaterializeRow()
   at System.Data.Entity.Core.Common.Internal.Materialization.Shaper`1.RowNestedResultEnumerator.MoveNext()
   at System.Data.Entity.Core.Common.Internal.Materialization.Shaper`1.ObjectQueryNestedEnumerator.TryReadToNextElement()
   at System.Data.Entity.Core.Common.Internal.Materialization.Shaper`1.ObjectQueryNestedEnumerator.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at Z.EntityFramework.Plus.QueryIncludeFilterProvider`1.Execute[TResult](Expression expression)
   at Orchestration.Managers.ProjectManager.GetProjectBudgets(Int32 projectId)
   at Presentation.Common.Controller.GetProjectBudgetConsumption(List`1 projectIds, Nullable`1 month, Nullable`1 year, Nullable`1 onlyValidatedHours)
   at Presentation.Controllers.DashboardController.GetProjectBudgetValues(Int32 projectId, Nullable`1 month, Nullable`1 year, Boolean onlyValidatedHours)
   at lambda_method(Closure , ControllerBase , Object[] )
   at System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters)
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.<BeginInvokeSynchronousActionMethod>b__39(IAsyncResult asyncResult, ActionInvocation innerInvokeState)
   at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult)
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethod(IAsyncResult asyncResult)
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3d()
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.AsyncInvocationWithFilters.<>c__DisplayClass46.<InvokeActionMethodFilterAsynchronouslyRecursive>b__3f()
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethodWithFilters(IAsyncResult asyncResult)
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.<>c__DisplayClass21.<>c__DisplayClass2b.<BeginInvokeAction>b__1c()
   at System.Web.Mvc.Async.AsyncControllerActionInvoker.<>c__DisplayClass21.<BeginInvokeAction>b__1e(IAsyncResult asyncResult)
```

Thanks for the attention 

Best Regards,

**Comment by JonathanMagnan** [comment]
Hello @Bladixx ,

Thank you for this additional information. Unfortunately, I'm not able to replicate the issue

Here is the code I made, perhaps you could use it to try to replicate your issue

```csharp
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Windows.Forms;

namespace Z.EntityFramework.Plus.Lab
{
    public partial class Form_Request_0774 : Form
    {
        public Form_Request_0774()
        {
            InitializeComponent();

            var id = 0;

            // CLEAR
            using (var ctx = new EntityContext())
            {
                ctx.Customers.RemoveRange(ctx.Customers);
                ctx.Projects.RemoveRange(ctx.Projects);
                ctx.SaveChanges();
            }

            // SEED
            using (var ctx = new EntityContext())
            {
                var customer1 = new Customer
                {
                    Name = "z1",
                    Description = "z1"
                };
                var customer2 = new Customer
                {
                    Name = "z2",
                    Description = "z2"
                };

                var project1 = new Project
                {
                    Name = "z1"
                };
                var project2 = new Project
                {
                    Name = "z2"
                };
                var project3 = new Project
                {
                    Name = "z3"
                };

                customer1.Projects = new List<Project>();
                customer1.Projects.Add(project1);
                customer1.Projects.Add(project2);
                customer2.Projects = new List<Project>();
                customer2.Projects.Add(project3);

                project1.ChildProjects = new List<Project>();
                project1.ChildProjects.Add(project2);
                project1.ChildProjects.Add(project3);

                ctx.Customers.Add(customer1);
                ctx.Customers.Add(customer2);
                ctx.SaveChanges();

                id = project1.Id;
            }

            // TEST
            using (var ctx = new EntityContext())
            {
                var project = ctx.Projects.AsQueryable()
                    .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted))
                    .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted).Select(t => t.Customer))
                    .IncludeFilter(x => x.Customer)
                    .FirstOrDefault(x => (x.Id == id) && !x.Deleted);
            }
        }

        public class EntityContext : DbContext
        {
            public EntityContext() : base("CodeFirstEntities")
            {
            }

            // ... context code ...
            public DbSet<Project> Projects { get; set; }
            public DbSet<Customer> Customers { get; set; }

            protected override void OnModelCreating(DbModelBuilder modelBuilder)
            {
                modelBuilder.Types().Configure(x => x.ToTable(GetType().DeclaringType != null ? GetType().DeclaringType.FullName.Replace(".", "_") + "_" + x.ClrType.Name : ""));

                base.OnModelCreating(modelBuilder);
            }
        }

        public class Project
        {
            public int Id { get; set; }
            public string Name { get; set; }
            public int? CustomerId { get; set; }
            public int? ParentId { get; set; }
            public virtual Customer Customer { get; set; }
            public virtual List<Project> ChildProjects { get; set; } = new List<Project>();
            public virtual Project Parent { get; set; }
            public bool Deleted { get; set; }
        }

        public class Customer
        {
            public int Id { get; set; }
            public string Name { get; set; }
            public string Description { get; set; }
            public virtual ICollection<Project> Projects { get; set; }
        }
    }
}
```

**Comment by OrestTkachuk** [comment]
I have same issue and I think you will be able to reproduce it in the following scenario:

- using(var ctx = new EntityContext())
 {
     var project = ctx.Projects.AsQueryable()
                        .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted))
                        .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted).Select(t => t.Customer))
                        .IncludeFilter(x => x.Customer)
                        .ToList();
}
-  using(var ctx = new EntityContext())
{
      var project = ctx.Projects.AsQueryable()
                        .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted))
                        .IncludeFilter(x => x.ChildProjects.Where(t => !t.Deleted).Select(t => t.Customer))
                       .IncludeFilter(x => x.Tasks)
                       .ToList();
}

And the error will be: 
{"The specified cast from a materialized 'System.Collections.Generic.IEnumerable`1[Tasks]' type to the 'Customer' type is not valid."}

**Comment by JonathanMagnan** [comment]
Hello @OrestTkachuk ,

Thank you for this additional information.

We will try this issue using a similar scenario as you.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Using the code you provided @OrestTkachuk , we successfully reproduced this issue.

Unfortunately, no quick fix has been found. We are still trying to investigate it to find if we can do something to solve this issue.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Nevermind,

We made a last test that gives us enough information to fix it.

The fix should be released very soon.

**Comment by OrestTkachuk** [comment]
Hello @JonathanMagnan,

I'm glad I was able to help.

**Comment by JonathanMagnan** [comment]
Hello @Bladixx , @OrestTkachuk 

The v1.6.7 has been released.

Let me know if you still have the issue.

Best Regards,

Jonathan

**Comment by dbpeterson2** [comment]
Can someone verify that v1.6.7 fixed this issue? I'm still getting the same error.

**Comment by OrestTkachuk** [comment]
Hello @JonathanMagnan 

The error doesn't appear anymore.

**Comment by JonathanMagnan** [comment]
Thank you @OrestTkachuk for confirming.

**Comment by JonathanMagnan** [comment]
Hello @dbpeterson2 ,

Do you think you can provide an example that still doesn't work?

Best Regards,

Jonathan

**Comment by dbpeterson2** [comment]
### The Code below does NOT work.

Project project =  (from p in ctx.Projects
                  .IncludeFilter(x => x.ChildProjects.Where(cp => cp.Status == 1))
                  .IncludeFilter(x => x.Customers)
                  .IncludeFilter(x => x.Orders.Where(o => o.Status = 1))
                 select p).FirstOrDefault();

Project project =  (from p in ctx.Projects
                  .IncludeFilter(x => x.Orders.Where(o => o.Status = 1))
                 select p).FirstOrDefault();

### ...but, this code DOES work.

Project project =  (from p in ctx.Projects
                  .IncludeFilter(x => x.Orders.Where(o => o.Status = 1))
                  .IncludeFilter(x => x.ChildProjects.Where(cp => cp.Status == 1))
                  .IncludeFilter(x => x.Customers)
                 select p).FirstOrDefault();

Project project =  (from p in ctx.Projects
                  .IncludeFilter(x => x.Orders.Where(o => o.Status = 1))
                 select p).FirstOrDefault();

---------------------------------------------------------------
Basically, it looks like the order of the .IncludeFilter() matters. This can cause problems if all queries are not ordered the same and means that I need to add the .IncludeFilter() for some items that I may not have wanted to have in the query.

### I also found that this does NOT work

Project project =  (from p in ctx.Projects
                  .IncludeFilter(x => x.Client)
                 select p).FirstOrDefault();

Project project =  (from p in ctx.Projects
                  .IncludeFilter(x => x.ChildProjects.Where(cp => cp.Status == 1))
                 select p).FirstOrDefault();

**Exception: The specified cast from a materialized 'System.Collections.Generic.IEnumerable`1[Domain.Objects.Entities.Project]' type to the 'Domain.Objects.Entities.Client' type is not valid.**

**Comment by JonathanMagnan** [comment]
Hello @dbpeterson2 ,

Thank you, we will investigate this case very soon.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Hello @dbpeterson2 ,

I didn't reproduced successfully the first example, are you using the latest version?

However, I successfully reproduced the second example that doesn't work. One line from our fix was missed.

The fix will be released Friday or Monday.

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Hello @dbpeterson2 

The v1.6.8 has been released, we have included a fix for the line missing that caused the Cast.

Please, reopen the issue if you still have the problem.

Best Regards,

Jonathan

**Comment by dbpeterson2** [comment]
@JonathanMagnan Thanks! Everything seems to be working now.
